File: AutomaticDrawing.Painter.BitmapAnimation/Painter/BitmapPainter.vb
Imports System.Drawing
Imports AutomaticDrawing.Core
''' <summary>
''' 提供位图动画的绘制器
''' </summary>
Public Class BitmapPainter
    Implements IPainter
    Public Event UpdatePaint As IPainter.UpdatePaintEventHandler Implements IPainter.UpdatePaint
    ''' <summary>
    ''' 视图
    ''' </summary>
    Public Property View As Bitmap
    ''' <summary>
    ''' 是否绘制原始数据
    ''' </summary>
    Public Property IsPaintRaw As Boolean

    Shared Rnd As New Random

    ''' <summary>
    ''' 创建并初始化一个实例
    ''' </summary>
    Public Sub New(view As Bitmap, Optional isPaintRaw As Boolean = True)
        Me.View = view
        Me.IsPaintRaw = isPaintRaw
    End Sub

    Public Sub Start(lines As List(Of ILine)) Implements IPainter.Start
        If IsPaintRaw Then
            PaintRaw(lines)
        Else
            PaintColorful(lines)
        End If
    End Sub

    Public Sub Pause() Implements IPainter.Pause
        Throw New NotImplementedException()
    End Sub
    Public Sub [Stop]() Implements IPainter.Stop
        Throw New NotImplementedException()
    End Sub

    Private Sub PaintRaw(lines As List(Of ILine))
        Using pg As Graphics = Graphics.FromImage(View)
            pg.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
            Dim totalCount As Integer = lines.SelectMany(Function(e As ILine) e.Vertices).Count
            Dim current As Integer = 0
            For Each SubSequence In lines
                For Each SubVertex In SubSequence.Vertices
                    Dim penSize As Single = SubVertex.Size
                    Dim brush As New SolidBrush(SubVertex.Color)
                    pg.FillRectangle(brush, New RectangleF(SubVertex.X - penSize / 2, SubVertex.Y - penSize / 2, penSize, penSize))
                    'pg.DrawLine(mypen, SubPoint, SubSequence.PointList.First)
                    current += 1
                    RaiseEvent UpdatePaint(Me, New UpdatePaintEventArgs(SubVertex, current / totalCount))
                Next
            Next
        End Using
    End Sub
    Private Sub PaintColorful(lines As List(Of ILine))
        Dim tempColor As Color = Color.FromArgb(255, 0, 0, 0)
        Using pg As Graphics = Graphics.FromImage(View)
            pg.SmoothingMode = Drawing2D.SmoothingMode.HighQuality
            Dim totalCount As Integer = lines.SelectMany(Function(e As ILine) e.Vertices).Count
            Dim current As Integer = 0
            For Each SubSequence In lines
                tempColor = Color.FromArgb(255, Rnd.NextDouble * 255, Rnd.NextDouble * 255, Rnd.NextDouble * 255)
                For Each SubPoint In SubSequence.Vertices
                    Dim penSize As Single = SubPoint.Size
                    Dim mypen As New Pen(tempColor, 1 + penSize)
                    pg.DrawEllipse(mypen, New RectangleF(SubPoint.X - penSize / 2, SubPoint.Y - penSize / 2, penSize, penSize))
                    'pg.DrawLine(mypen, SubPoint, SubSequence.PointList.First)
                    current += 1
                    RaiseEvent UpdatePaint(Me, New UpdatePaintEventArgs(SubPoint, current / totalCount))
                Next
            Next
        End Using
    End Sub

End Class

File: AutoPaint.Painter.BitmapAnimation/Painter/BitmapPainter.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using AutoPaint.Core;
using System.Drawing;

namespace AutoPaint.Painter
{
    public class BitmapPainter : IPainter
    {
        public event EventHandler<OnPaintingUpdatedEventArgs> OnPaintingUpdated;
        public Bitmap SourceBitmap { get; set; }

        public BitmapPainter(Bitmap targetBitmap)
        {
            this.SourceBitmap = targetBitmap;
        }

        public void Start(IPainting painting)
        {
            PaintLines(painting.Lines);
        }

        public void Pause()
        {
            throw new NotImplementedException();
        }
        public void Resume()
        {
            throw new NotImplementedException();
        }
        public void Stop()
        {
            throw new NotImplementedException();
        }

        private void PaintLines(IEnumerable<ILine> lines)
        {
            using (Graphics pg = Graphics.FromImage(SourceBitmap))
            {
                pg.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
                int totalCount = lines.Count();
                int current = 0;
                foreach (Line line in lines)
                {
                    if (line.IsOutline)
                    {
                        PaintOutline(pg, line);
                    }
                    else
                    {
                        PaintLine(pg, line);
                    }
                    current += 1;
                    OnPaintingUpdated?.Invoke(this, new OnPaintingUpdatedEventArgs(line.Vertices.FirstOrDefault(), current / (float)totalCount));
                }
            }
        }

        private static void PaintLine(Graphics pg, ILine line)
        {
            foreach (var vertex in line.Vertices)
            {
                float penSize = vertex.Size;
                float halfSize = penSize / 2;
                SolidBrush brush = new SolidBrush(vertex.Color);
                pg.FillRectangle(brush, new RectangleF(vertex.X - halfSize, vertex.Y - halfSize, penSize, penSize));
            }
        }

        private static void PaintOutline(Graphics pg, ILine line)
        {
            var vertex = line.Vertices[0];
            float penSize = 1;
            //float penSize = vertex.Size;
            SolidBrush brush = new SolidBrush(vertex.Color);
            var points = line.Vertices.Select(v => new PointF(v.X, v.Y)).ToArray();
            pg.DrawLines(new Pen(brush, penSize), points);
        }
    }
}
